File: nlpcraft-examples/solarsystem/src/main/java/org/apache/nlpcraft/examples/solarsystem/loaders/SolarSystemDiscoversValueLoader.java
package org.apache.nlpcraft.examples.solarsystem.loaders;

import org.apache.nlpcraft.examples.solarsystem.api.SolarSystemOpenApiService;
import org.apache.nlpcraft.model.NCElement;
import org.apache.nlpcraft.model.NCValue;
import org.apache.nlpcraft.model.NCValueLoader;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

public class SolarSystemDiscoversValueLoader implements NCValueLoader {
    private SolarSystemOpenApiService api;

    public SolarSystemDiscoversValueLoader(SolarSystemOpenApiService api) {
        this.api = api;
    }

    private NCValue mkValue(String discInfo) {
        List<String> syns = new ArrayList<>();

        Arrays.stream(discInfo.split(" ")).map(p -> p.strip()).filter(p -> !p.isEmpty()).forEach(p -> {
            String lc = p.toLowerCase();

            syns.add(lc);

            int lastNameIdx = lc.lastIndexOf(" ");

            // Tries to detect last name.
            if (lastNameIdx > 0)
                syns.add(lc.substring(lastNameIdx + 1));

        });

        return new NCValue() {
            @Override
            public String getName() {
                return discInfo;
            }

            @Override
            public List<String> getSynonyms() {
                return syns;
            }
        };
    }

    @Override
    public Set<NCValue> load(NCElement owner) {
        return api.getAllDiscovers().stream().map(this::mkValue).collect(Collectors.toSet());
    }
}

File: nlpcraft-examples/solarsystem/src/main/java/org/apache/nlpcraft/examples/solarsystem/loaders/SolarSystemDiscoversValueLoader.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.nlpcraft.examples.solarsystem.loaders

import org.apache.nlpcraft.examples.solarsystem.api.SolarSystemOpenApiService
import org.apache.nlpcraft.model.{NCElement, NCValue, NCValueLoader}

import java.util
import scala.collection.mutable.ArrayBuffer
import scala.jdk.CollectionConverters.{SeqHasAsJava, SetHasAsJava}

class SolarSystemDiscoversValueLoader extends NCValueLoader {
    // Example: "Scott S. Sheppard, David C. Jewitt"
    private def mkValue(discInfo: String): NCValue = {
        val syns = ArrayBuffer.empty[String]

        for (d <- discInfo.split(",").map(_.trim).filter(_.nonEmpty)) {
            val lc = d.toLowerCase

            syns += lc

            val lastNameIdx = lc.lastIndexOf(" ")

            // Tries to detect last name.
            if (lastNameIdx > 0)
                syns += lc.substring(lastNameIdx + 1)
        }

        new NCValue {
            override def getName: String = discInfo
            override def getSynonyms: util.List[String] = syns.asJava
        }
    }

    override def load(owner: NCElement): util.Set[NCValue] =
        SolarSystemOpenApiService.getInstance().getAllDiscovers.map(mkValue).toSet.asJava
}
